streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -3523,24 +3523,10 @@ final class ZStream[-R, +E, +A] private (val channel: ZChannel[R, Any, Any, Any,
    * Converts this stream into a `scala.collection.Iterator` wrapped in a scoped
    * [[ZIO]]. The returned iterator will only be valid within the scope.
    */
-  def toIterator(implicit trace: Trace): ZIO[R with Scope, Nothing, Iterator[Either[E, A]]] =
-    for {
-      runtime <- ZIO.runtime[R]
-      pull    <- toPull
-    } yield {
-      def unfoldPull: Iterator[Either[E, A]] =
-        runtime.unsafe.run(pull)(trace, Unsafe.unsafe) match {
-          case Exit.Success(chunk) => chunk.iterator.map(Right(_)) ++ unfoldPull
-          case Exit.Failure(cause) =>
-            cause.failureOrCause match {
-              case Left(None)    => Iterator.empty
-              case Left(Some(e)) => Iterator.single(Left(e))
-              case Right(c)      => throw FiberFailure(c)
-            }
-        }
-
-      unfoldPull
-    }
+  def toIterator(implicit trace: Trace): ZIO[R with Scope, Nothing, Iterator[Either[E, A]]] = for {
+    runtime <- ZIO.runtime[R]
+    pull    <- either.toPull
+  } yield unfoldPull(runtime, pull)(trace, Unsafe.unsafe).flatten
```

Comment:
I have a feeling that this will destroy chunking, or at the very least cause each element of the stream to be evaluated in a new call to `runtime.unsafe.run`. This will be very expensive as each time we'd need to create a new Fiber to evaluate it. If we can't reuse `unfoldPull` I'd say it's probably better to specialize this implementation instead

Reply:
`pull` returns a `Chunk`, that's why we call `.flatten`

Reply:
Aghh I see